### PR TITLE
issue state ping pong

### LINKED ISSUE #22: Bug: fix reopening issues
Check this [issue](https://github.com/ubiquity/devpool-directory/issues/984#event-12112923776). You may see that it is constantly opened and closed on every [sync workflow](https://github.com/ubiquity/devpool-directory/blob/development/.github/workflows/sync-issues.yml) run.

What should be done:
- find the root cause of why the issue is reopened
- fix the bug

### PR DESCRIPTION
**Pull request by Keyrxng**
Resolves https://github.com/ubiquity/devpool-directory-bounties/issues/22

- refactored the logic a bit
- only two scenarios should reopen a closed devpool bounty

QA

- https://github.com/Keyrxng/devpool-directory/actions/runs/8296972473/job/22707048890 (most recent)
- https://github.com/Keyrxng/devpool-directory/actions/runs/8296652112 5x runs
- my devpool stays at 21 tasks consistently
- I'm just waiting for the devpool workflow to run again and spike up for one last validation

### CODE REVIEW

**Comment by Keyrxng** [comment]
@rndquu good to go mate

devpool just went from 20 to 29 while mine went to 20 as I started this issue so it's unavailable now

##### Logs
```
Closed (assigned-open)::
Devpool Issue: https://github.com/Keyrxng/devpool-directory/issues/63
Project Issue: https://github.com/ubiquity/devpool-directory-bounties/issues/22
```

File: total-rewards.json
Comment (0x4007) [inline total-rewards.json]: That doesn't seem right 0 for assigned.

File: total-rewards.json
Comment (0x4007) [inline total-rewards.json]: This seems like too many assigned! It's a separate issue!

Comment (Keyrxng) [reply total-rewards.json]: from the current dev branch, those small numbers are from my devpool.

Bit of a nightmare working with that from a forked perspective imo, manageable but tedious

```ts
{
  "rewards": {
    "notAssigned": 2850,
    "assigned": 825,
    "completed": 69593,
    "total": 73268
  },
  "tasks": {
    "notAssigned": 312,
    "assigned": 285,
    "completed": 354,
    "total": 597
  }
}
```

Comment (0x4007) [reply total-rewards.json]: Interesting stats. Big numbers are always good for marketing.

Comment (Keyrxng) [reply total-rewards.json]: you know what they say about big numbers

Comment (0x4007) [reply total-rewards.json]: Big funding

**Review by rndquu — APPROVED** [review]

**Review by rndquu — CHANGES_REQUESTED** [review]
Testing workflow is [failing](https://github.com/ubiquity/devpool-directory/actions/runs/8343918622/job/22835140783?pr=1145)

**Comment by Keyrxng** [comment]
> Testing workflow is [failing](https://github.com/ubiquity/devpool-directory/actions/runs/8343918622/job/22835140783?pr=1145)

It's failing in error I believe, can you try running the workflow again?

```yml
● Devpool Updates › Close DevPool issues with no price

    expect(received).toMatchObject(expected)

    - Expected  - 4
    + Received  + 4
```

**Comment by rndquu** [comment]
> > Testing workflow is [failing](https://github.com/ubiquity/devpool-directory/actions/runs/8343918622/job/22835140783?pr=1145)
> 
> It's failing in error I believe, can you try running the workflow again?
> 
> ```yaml
> ● Devpool Updates › Close DevPool issues with no price
> 
>     expect(received).toMatchObject(expected)
> 
>     - Expected  - 4
>     + Received  + 4
> ```

Rerun it, the [error](https://github.com/ubiquity/devpool-directory/actions/runs/8343918622/job/22874312148) persists:
```
    @@ -22,24 +22,24 @@
            "subscriptions_url": "",
            "type": "",
            "url": "",
          },
          "author_association": "NONE",
    -     "body": "body",
    +     "body": "https://github.com/ubiquity/test-repo/issues/1",
          "closed_at": null,
          "comments": 0,
          "comments_url": "",
          "created_at": "",
          "events_url": "",
          "html_url": "https://github.com/ubiquity/devpool-directory/issues/1",
          "id": 1,
          "labels": Array [
            Object {
    -         "name": "Pricing: 200 USD",
    +         "name": "Pricing: not set",
            },
            Object {
    -         "name": "Time: 1h",
    +         "name": "Partner: ubiquity/test-repo",
            },
            Object {
              "name": "id: 1",
            },
          ],
```

**Comment by 0x4007** [comment]
> > > Testing workflow is [failing](https://github.com/ubiquity/devpool-directory/actions/runs/8343918622/job/22835140783?pr=1145)
> 
> > 
> 
> > It's failing in error I believe, can you try running the workflow again?
> 
> > 
> 
> > ```yaml
> 
> > ● Devpool Updates › Close DevPool issues with no price
> 
> > 
> 
> >     expect(received).toMatchObject(expected)
> 
> > 
> 
> >     - Expected  - 4
> 
> >     + Received  + 4
> 
> > ```
> 
> 
> 
> Rerun it, the [error](https://github.com/ubiquity/devpool-directory/actions/runs/8343918622/job/22874312148) persists:
> 
> ```
> 
>     @@ -22,24 +22,24 @@
> 
>             "subscriptions_url": "",
> 
>             "type": "",
> 
>             "url": "",
> 
>           },
> 
>           "author_association": "NONE",
> 
>     -     "body": "body",
> 
>     +     "body": "https://github.com/ubiquity/test-repo/issues/1",
> 
>           "closed_at": null,
> 
>           "comments": 0,
> 
>           "comments_url": "",
> 
>           "created_at": "",
> 
>           "events_url": "",
> 
>           "html_url": "https://github.com/ubiquity/devpool-directory/issues/1",
> 
>           "id": 1,
> 
>           "labels": Array [
> 
>             Object {
> 
>     -         "name": "Pricing: 200 USD",
> 
>     +         "name": "Pricing: not set",
> 
>             },
> 
>             Object {
> 
>     -         "name": "Time: 1h",
> 
>     +         "name": "Partner: ubiquity/test-repo",
> 
>             },
> 
>             Object {
> 
>               "name": "id: 1",
> 
>             },
> 
>           ],
> 
> ```
> 
> 

I think the test just might be poorly written. It seems to be checking that price is a specific value etc when really it should be a loose match (every issue has different labels!) @FernandVEYRIER

**Comment by gentlementlegen** [comment]
@pavlovcik the test results should be an exact snapshot of what is expected to be answered by the API, otherwise there could be values that change and errors we could not foresee. But I don't understand why you say it is failing, it seems successful on the latest run https://github.com/ubiquity/devpool-directory/actions/runs/8358643627/job/22880364593?pr=1145 and to make sure I pulled the branch and tested on my own repo with the latest commit which also ran fine https://github.com/FernandVEYRIER/devpool-directory/actions/runs/8368745527/job/22913337163

**Comment by 0x4007** [comment]
I'm just going off of rndquu comment

I thought this throws an error in the test 

```diff

-     "body": "body",
+     "body": "https://github.com/ubiquity/test-repo/issues/1",

```

But yes I see there was a commit after my comment that passes checks.

**Comment by gitcoindev** [comment]
> I'm just going off of rndquu comment
> 
> I thought this throws an error in the test
> 
> ```diff
> -     "body": "body",
> +     "body": "https://github.com/ubiquity/test-repo/issues/1",
> ```
> 
> But yes I see there was a commit after my comment that passes checks.

The merge conflicts should be resolved then and ready to be merged or is there still anything missing?

**Comment by 0x4007** [comment]
`total-rewards.json` probably will also be a headache for future merge conflicts- we should add to .gitignore?

File: tests/main.test.ts
Comment (0x4007) [inline tests/main.test.ts]: Why are you combining opt out if those are explicitly meant to be excluded

Comment (Keyrxng) [reply tests/main.test.ts]: https://github.com/ubiquity/devpool-directory/commit/025868c32c5571f6c3457f46d1adce3b1d7aa8cf

It comes from this commit, I must have made a slight mistake in how I've merged it that should be eresdev's commit

Comment (0x4007) [reply tests/main.test.ts]: @EresDev rfc

Comment (EresDev) [reply tests/main.test.ts]: > @EresDev rfc

This part had multiple problems and I tried to address some with this fix I didn't want to put more time into this because the work I was doing was unrelated to this. But let's see what is needed now. 

- before you use in, and exclude out, this part needs to deal with out/in with a single value. It was treating properly the fully qualified repository name "GithubAccountName/RepostioryName", but if it was just "GithubAccountName", it treated it as "RepostioryName" That is why I had to put a check to only use a fully qualified repository name. This needs to be fixed. https://github.com/ubiquity/devpool-directory/blob/131e16d78d245c1f4bf03e31b6935ad78900d450/tests/main.test.ts#L28


-  after the above fix, use the in, and exclude the out. 
- also move the mocks folder out of the root directory into the test folder. mocks are only used by tests so they should live together.

Comment (EresDev) [reply tests/main.test.ts]: One suggestion is always to use a full repository name in opt.json "GithubAccountName/RepostioryName", and it will make out/in redundant. This will result in lesser & cleaner code, less requests, and clear intent. But it is just a suggestion as this will require redoing a few parts and I was not the part of initial discussion on this. But @pavlovcik has the final say on this.

Comment (0x4007) [reply tests/main.test.ts]: No it definitely needs to be able to automatically include new repositories for privileged organizations like @ubiquity. 

This means that we must support org only syntax.

Comment (gitcoindev) [reply tests/main.test.ts]: I have not looked in this PR for quite some time, but receiving a dozen notification emails a day about closing and reopening tasks I feel I will join the reviewers party.

About opt in and opt out, I initially implemented quite a simple solution at https://github.com/ubiquity/devpool-directory/pull/982/files that worked correctly at least to extent to the tests coverage at that time. Perhaps this can be refactored but it did seem to work, first by adding repositories, then deleting opt outs from the set. I did not follow up the recent changes, what was the main issue with the previous approach?

Comment (0x4007) [reply tests/main.test.ts]: > I have not looked in this PR for quite some time, but receiving a dozen notification emails a day about closing and reopening tasks I feel I will join the reviewers party.

Unless you're very partial towards using email, I highly recommend you use github.com/notifications ! This is because you only need to "pull" (instead of push) the latest notification in each thread. Emails give you multiple "push" notifications for every activity in every thread. So you have a significant higher amount of clutter to sift through. 

> About opt in and opt out, I initially implemented quite a simple solution at https://github.com/ubiquity/devpool-directory/pull/982/files that worked correctly at least to extent to the tests coverage at that time. Perhaps this can be refactored but it did seem to work, first by adding repositories, then deleting opt outs from the set. I did not follow up the recent changes, what was the main issue with the previous approach?

I don't know maybe somebody else can chime in.

Comment (gitcoindev) [reply tests/main.test.ts]: > > I have not looked in this PR for quite some time, but receiving a dozen notification emails a day about closing and reopening tasks I feel I will join the reviewers party.
> 
> Unless you're very partial towards using email, I highly recommend you use github.com/notifications ! This is because you only need to "pull" (instead of push) the latest notification in each thread. Emails give you multiple "push" notifications for every activity in every thread. So you have a significant higher amount of clutter to sift through.
> 

The email was good enough till recently. Great tip with the notifications, I will check this out asap!

**Review by rndquu — CHANGES_REQUESTED** [review]
Somehow the code in this PR duplicates issues with the same id although the expected behaviour is to have exactly 1 issue:
Screenshot 2024-03-25 at 17 11 46


Pls fix

**Comment by Keyrxng** [comment]
- https://github.com/Keyrxng/devpool-directory/actions/runs/8432637740/job/23092097971 no changes
- https://github.com/Keyrxng/devpool-directory/actions/runs/8432656790/job/23092157309 reopening closed

> Pls fix

I made a mess of the merge and ended up with 4 for-loops, merged properly and split it into two functions. All tests passing, QA runs linked.

**Review by rndquu — APPROVED** [review]
Works fine, @Keyrxng thank you

**Comment by 0x4007** [comment]
> Works fine, @Keyrxng thank you

Perhaps we should merge this if everything checks out?

**Pull request merged**
